---
title: "findPvalueCutoff"
shorttitle: "Select Differentially Expressed Genes At False Positive Rate Less Than Five Percent"
author:
- name: "Vijaykumar Yogesh Muley"
  affiliation: 
  - Instituto de Neurobiología, Universidad Nacional Autónoma de México, Querétaro, Mexico
  email: vijay.muley@gmail.com
date: "`r Sys.Date()`"
output: 
    BiocStyle::html_document:
      toc: true
      toc_depth: 2
bibliography: `'r system.file("inst/REFERENCES.bib", package="findPvalueCutoff")`'`
vignette: >
    %\VignetteIndexEntry{Vignette Title}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

The findPvalueCutoff package is intended to use with p-values resulting from differential gene expression analysis tools to estimate a data-driven p-value cutoff. Conventionally, adjusted p-value cutoff of 0.05 or less is used to selected differentially expressed genes. However, there is a great debate on whether 0.05 p-value cutoff should be used [@Amrhein2018]. Sample size of the groups compared affect the generated p-values and hence conventional p-value cutoff may not be appropriate. The advantage with using findPvalueCutoff is that the p-value cutoff is predicted based on distribution of p-values itself by randomizing their order using sophisticated simulation based permutation resampling. This data-driven approach estimates p-value cutoff where one can expected five percent false positive error rate in detecting differentially expressed genes. In the subsequent text, the details are presented on how to compute false discovery rate at various p-value cutoff, and selection of differentially expressed genes.

## Contact

For questions regarding findPvalueCutoff, contact the author at vijay.muley\@gmail.com

# INSTALLATION

You need devtools package to install findPvalueCutoff. This can be done by installing devtools by typing install.packages(devtools) command in R. If devtools package is already installed then you can use following command to load devtools package in current R environment.  

```{r, eval=FALSE}
library(devtools)

```

Then you can copy and paste following command to install findPvalueCutoff package from github repository.

```{r, eval=FALSE}
devtools::install_github("vijaykumarmuley/findPvalueCutoff", build_vignettes = TRUE)

```

Once installed you can load findPvalueCutoff into your workspace:


```{r, eval=TRUE}
library(findPvalueCutoff)
```


Note that this command has to be run every time you restart R in order to be able to use the
package. To get immediate help on findPvalueCutoff, type:


```{r}
?findPvalueCutoff
```

findPvalueCutoff function needs a vector of uncorrected p-values resulting from differential gene expression analysis tools. Some parameters need to be adjusted and details are giving below. 

# Estimating false positive rates at various p-value cutoffs

I will use example data provided with the package, which contains output similar to generated by differential gene expression analysis tools such as limma, edgeR or DEGseq. The following command will load the data which is a data.frame named pvalue


```{r}
data(pvalue)
```


To check first few lines of the data, you can type following commands


```{r}
head(pvalue)
```

There are two ways to run findPvalueCutoff. 1) use a vector of predefined p-value cutoffs (manually selected set) or 2) set a lower and upper range and choose a interval so that this function generates a series of p-value cutoffs from lower to upper bound by incrementing lower bound with desired number that provided using interval parameter. I have given examples of both ways below.

## Using predefined p-value thresholds

findPvalueCutoff function generates random samples of the input p-values. Hence, the output would be little bit different with each run on the same data. Therefore, I first set a seed to a random number using set.seed function to reproduce the same results. Then I run findPvalueCutoff by setting 

1. x paramater to differential expression values from the column P.Value in the pvalue data.frame, 
2. thresholds parameter to manually selected p-value cutoffs at which I would like to compute false positive rates, 
3. nsims parameter to 100 to create 100 randomized datasets from the given p-values

nsims should be set to at least 1000 when performing analysis using real data. The recommended value is 10000 for accurate results.

After setting these parameters, I run findPvalueCutoff function and store the output in fdr object.

```{r}
set.seed(1979)
fdr <- findPvalueCutoff(x = pvalue$P.Value, thresholds = c(0.0001, 0.001, 0.01, 0.1, 0.15, 0.2), nsims = 100)
```

As shown below fdr object is a numeric vector containing false positive rates in percentages computed at a given series of p-value cutoffs, which are names of the elements in the fdr object. For instance, approximately 5 percent of the differentially expressed genes are likely to be falsly detected by chance along at the p-value cutoff of 0.01.

```{r}
fdr
```

Lets compare how many genes are differentially expressed using conventional cutoff of 0.05 after adjusted p-values, and the 0.001 cutoff on unadjusted pvalues suggested by findPvalueCutoff function.

```{r}
sum(pvalue$adj.P.Val<=0.05)
sum(pvalue$P.Value<=0.001)

```

As shown above, `r sum(pvalue$adj.P.Val<=0.05)` genes were identified as differentially expressed using conventional p-value cutoff, while only `r sum(pvalue$P.Value<=0.001)` genes are differentially expressed according to p-value cutoff estimated by findPvalueCutoff function. The expected detection error rate at this cutoff is close to five percent of the total identified genes. 


These results can be ploted as follow, and can be saved later use.

```{r, fig.cap = "False positive error rates predicted from random datasets according to the uncorrected differential expression p-values. The false positive error rate represents the ratio of the number of significant p-values from the randomized dataset (control experiment) to the number of significant p-values from the real dataset below a certain p-value. False positive error rate is close to 5 percent at p-value cutoff of 0.01."}
barplot(fdr, names = names(fdr), col =  "black", border = "white", xlab = "P-value cutoff", ylab = "False Positive Rate (in percent)")
```

The plot shows false positive rate is close to five at 0.001 p-value cutoff. This should be used for selecting differentially expressed genes.

## Using a range of pvalues at specific interval

findPvalueCutoff provide two more parameters to quickly get an idea about false positive rates at a series of p-value cutoff. This can be done by running it with following parameters. 

It is also possible to provide a range of p-values with an interval to get an idea of p-value cutoff where it is more likely to find false positive rate as low as possible. To do that, you can choose a lower and upper bound of the p-value cutoff, and increment lower bound by a interval to get a series of p-value cutoffs within those bounds. Here I choose p-values lower bound of 0 to upper bound of 0.2, and increment by 0.02.

1. x paramater to differential expression values from the column P.Value in the pvalue data.frame, 
2. range parameter to provide a lower and upper bound of the p-values, 
3. interval parameter to a number to increment lower bound of the p-value progressively until upper bound provided with range parameter,
4. nsims parameter to 100 to create 100 randomized datasets from the given p-values

In this example, I set range between 0 to 0.2 with interval 0.02, and excute the function as shown below,

```{r}
set.seed(1979)
fdr <- findPvalueCutoff(x = pvalue$P.Value, range = c(0, 0.2), interval = 0.02, nsims = 100)
```

These results can be ploted as follow
```{r, fig.cap = "False positive error rates predicted from random datasets according to the uncorrected differential expression p-values. The false positive error rate represents the ratio of the number of significant p-values from the randomized dataset (control experiment) to the number of significant p-values from the real dataset below a certain p-value. This plot gives an idea that false positive error rate is five perent or less between p-values 0 to 0.02. One can now use p-values between this range to get more finer false positive rate and corresponding p-value cutoff."}
barplot(fdr, names = names(fdr), col =  "black", border = "white", xlab = "P-value cutoff", ylab = "False Positive Rate (in percent)")
```

Plot shows that false positive rate of 5 is estimted between the range of 0 and 0.02 p-value cutoffs. This range can be used to get more finer false positive rate and the corresponding p-values.


```{r}
sessionInfo()
```
